### PR TITLE
Add ForbiddenResult

### LINKED ISSUE #3242: Add a new ForbiddenResult
This would be almost identical to ChallengeResult, except it calls Forbid instead of Challenge, which would make it easier to directly specify access denied flows (i.e. signal cookies to redirect to the access denied page).

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #3242

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @rynowak \ @HaoK

File: src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        public ForbiddenResult(string authenticationScheme)
+            : this(new[] { authenticationScheme })
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ForbiddenResult"/> with the
+        /// specified authentication schemes.
+        /// </summary>
+        public ForbiddenResult(IList<string> authenticationSchemes)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: add param docs

File: src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        public ForbiddenResult()
+            : this(new string[] { })
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ForbiddenResult"/> with the
+        /// specified authentication scheme.
+        /// </summary>
+        public ForbiddenResult(string authenticationScheme)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: add param docs

File: src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        public ForbiddenResult(IList<string> authenticationSchemes)
+            : this(authenticationSchemes, properties: null)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ForbiddenResult"/> with the
+        /// specified <paramref name="properties"/>.
+        /// </summary>
+        public ForbiddenResult(AuthenticationProperties properties)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: add param docs

File: src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            : this(new string[] { }, properties)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ForbiddenResult"/> with the
+        /// specified authentication scheme and <paramref name="properties"/>.
+        /// </summary>
+        public ForbiddenResult(string authenticationScheme, AuthenticationProperties properties)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: add param docs

File: src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            : this(new[] { authenticationScheme }, properties)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="ForbiddenResult"/> with the
+        /// specified authentication schemes and <paramref name="properties"/>.
+        /// </summary>
+        public ForbiddenResult(IList<string> authenticationSchemes, AuthenticationProperties properties)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: add param docs

File: src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        {
+            AuthenticationSchemes = authenticationSchemes;
+            Properties = properties;
+        }
+
+        /// <summary>
+        /// The list of authentication components that should handle the authentication challenge
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: Gets or sets (all of em)

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: should this say `authentication schemes`? Is there anything easy we can put here to say what this means?

File: src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var loggerFactory = context.HttpContext.RequestServices.GetRequiredService<ILoggerFactory>();
+            var logger = loggerFactory.CreateLogger<ForbiddenResult>();
+
+            var authentication = context.HttpContext.Authentication;
+
+            if (AuthenticationSchemes != null && AuthenticationSchemes.Count > 0)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: Count here isn't needed.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: `AuthenticationSchemes?.Count > 0` #PranavPoints :registered:

File: src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    public static class ForbiddenResultLoggerExtensions
+    {
+        private static readonly Action<ILogger, string[], Exception> _resultExecuting =
+            LoggerMessage.Define<string[]>(
+                LogLevel.Information,
+                eventId: 1,
+                formatString: $"Executing {nameof(ForbiddenResult)} with authentication schemes ({{Schemes}}).");
+
+        public static void ForbiddenResultExecuting(this ILogger logger, IList<string> authenticationSchemes)
+            => _resultExecuting(logger, authenticationSchemes.ToArray(), null);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs]: This is really ugly. just use the objectively better normal syntax.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs]: named param for `null`

File: src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+using System.Linq;
+using Microsoft.AspNet.Mvc.Core;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Mvc.Logging
+{
+    /// <summary>
+    /// Extensions methods for logging <see cref="ForbiddenResult"/> instances.
+    /// </summary>
+    public static class ForbiddenResultLoggerExtensions
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs]: internal. Yes, really.

File: src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.Mvc.Core;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs]: What's this for?

File: src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs (hunk 9 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs]: No need for docs on these FYI

File: test/Microsoft.AspNet.Mvc.Core.Test/ForbiddenResultTest.cs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+                routeData,
+                new ActionDescriptor());
+
+            // Act
+            await result.ExecuteResultAsync(actionContext);
+
+            // Assert
+            authenticationManager.Verify();
+        }
+
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ForbiddenResultTest.cs]: Would like to see a case with an empty list of schemes. I almost gave the feedback to oversimplify that loop, so it seems worth having.

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs (hunk 6 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: @Tratcher was the one that came up with the term authentication components. Schemes seems a bit easier to follow

File: src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs (hunk 7 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: Well I care about unnecessary boxing.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ForbiddenResult.cs]: :facepunch: - there's some unnecessary boxing for you.

File: src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs (hunk 8 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs]: > named param for null

Go away :trollface:

File: src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs (hunk 10 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs]: Removed

**Comment by pranavkm** [comment]
:up: :date:

File: src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs (hunk 8 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Logging/ForbiddenResultLoggerExtensions.cs]: http://data.whicdn.com/images/56669995/original.jpg

**Comment by rynowak** [comment]
:shipit:

**Pull request closed without merging**